activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -29,7 +29,14 @@ def scope
         end
 
         def records_for(ids)
-          query_scope(ids)
+          records = query_scope(ids)
+          # Respect the order on `scope` if it exists, else use the order `ids`,
+          # which is already sorted based on the order of the through association
+          if scope.values[:order].any?
+            records
+          else
+            records.sort_by { |record| ids.index record[association_key_name] }
```

Comment:
I think it's going to be one of those situations where it depends on number of records in memory and size of the table since it'll mean joining the through association table to the destination table for the ordering to occur. We should also look at what happens when you have a nesting of more than one level - if we're preloading a nested through association we could end up with an expensive query.